File: customer-address-jpa/src/test/java/com/izertis/example/service/CustomerServiceTest.java
package com.izertis.example.service;

import com.izertis.example.config.ServicesInMemoryConfig;
import com.izertis.example.domain.Address;
import com.izertis.example.domain.Customer;
import com.izertis.example.service.dtos.CustomerSearchCriteria;
import com.izertis.example.repository.jpa.inmemory.CustomerRepositoryInMemory;
import com.izertis.example.service.impl.CustomerServiceImpl;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.PageRequest;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/** Acceptance Test for CustomerService. */
class CustomerServiceTest {

    private final Logger log = LoggerFactory.getLogger(getClass());

    ServicesInMemoryConfig context = new ServicesInMemoryConfig();

    CustomerServiceImpl customerService = context.customerService();

    CustomerRepositoryInMemory customerRepository = context.customerRepository();

    @BeforeEach
    void setUp() {
        context.reloadTestData();
    }

    @Test
    void createCustomerTest() {
        var input = new Customer();
         input.setName("name");
         input.setEmail("me@email.com");
         input.setAddresses(List.of(new Address() //
                 .setStreet("street")
                 .setCity("city")
         ));
        var customer = customerService.createCustomer(input);
        assertNotNull(customer.getId());
        assertTrue(customerRepository.containsEntity(customer));
    }

    @Test
    void getCustomerTest() {
        var id = 1L; // TODO fill id
        var customer = customerService.getCustomer(id);
        assertTrue(customer.isPresent());
    }

    @Test
    void updateCustomerTest() {
        var id = 1L; // TODO fill id
        var input = new Customer();
        input.setName("name");
        input.setEmail("me@email.com");
        input.setAddresses(List.of(new Address() //
                .setStreet("street")
                .setCity("city")
        ));
        assertTrue(customerRepository.containsKey(id));
        var customer = customerService.updateCustomer(id, input);
        assertTrue(customer.isPresent());
        assertTrue(customerRepository.containsEntity(customer.get()));
    }

    @Test
    void deleteCustomerTest() {
        var id = 1L; // TODO fill id
        assertTrue(customerRepository.containsKey(id));
        customerService.deleteCustomer(id);
        assertFalse(customerRepository.containsKey(id));
    }

    @Test
    void searchCustomersTest() {
        var searchCriteria = new CustomerSearchCriteria();
        var results = customerService.searchCustomers(searchCriteria, PageRequest.of(0, 10));
        Assertions.assertNotNull(results);
        Assertions.assertFalse(results.isEmpty());
    }

}

File: kustomer-address-jpa/src/test/kotlin/com/izertis/example/service/impl/CustomerServiceTest.kt
package com.izertis.example.service.impl

import com.izertis.example.config.ServicesInMemoryConfig
import com.izertis.example.domain.Address
import com.izertis.example.domain.Customer
import com.izertis.example.repository.jpa.inmemory.CustomerRepositoryInMemory
import com.izertis.example.service.CustomerService
import com.izertis.example.service.dtos.CustomerSearchCriteria
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest

/** Acceptance Test for CustomerService. */
class CustomerServiceTest {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    private val context = ServicesInMemoryConfig()
    private val customerService: CustomerService = context.customerService()
    private val customerRepository: CustomerRepositoryInMemory = context.customerRepository()

    @BeforeEach
    fun setUp() {
        context.reloadTestData()
    }

    @Test
    fun createCustomerTest() {
        val input = Customer().apply {
            name = "name"
            email = "me@email.com"
            addresses = mutableListOf(Address().apply {
                street = "street"
                city = "city"
            })
        }

        val customer = customerService.createCustomer(input)
        assertNotNull(customer.id)
        assertTrue(customerRepository.containsEntity(customer))
    }

    @Test
    fun getCustomerTest() {
        val id = 1L
        val customer = customerService.getCustomer(id)
        assertNotNull(customer)
    }

    @Test
    fun updateCustomerTest() {
        val id = 1L
        val input = Customer().apply {
            name = "name"
            email = "me@email.com"
            addresses = mutableListOf(Address().apply {
                street = "street"
                city = "city"
            })
        }

        assertTrue(customerRepository.containsKey(id))
        val customer = customerService.updateCustomer(id, input)
        assertNotNull(customer)
        assertTrue(customerRepository.containsEntity(customer!!))
    }

    @Test
    fun deleteCustomerTest() {
        val id = 1L
        assertTrue(customerRepository.containsKey(id))
        customerService.deleteCustomer(id)
        assertFalse(customerRepository.containsKey(id))
    }

    @Test
    fun searchCustomersTest() {
        val searchCriteria = CustomerSearchCriteria()
        val results = customerService.searchCustomers(searchCriteria, PageRequest.of(0, 10))
        assertNotNull(results)
        assertFalse(results.isEmpty)
    }
}
